compiler/lcalc/closure_conversion.ml

```diff
@@ -324,16 +299,277 @@ let closure_conversion (p : 'm program) : 'm program Bindlib.box =
             }
           in
           let _free_vars, new_expr = transform_closures_expr ctx expr in
-          ( (Var.Set.add var toplevel_vars, decl_ctx),
+          ( Var.Set.add var toplevel_vars,
             Bindlib.box_apply
               (fun e -> Topdef (name, ty, e))
               (Expr.Box.lift new_expr) ))
       ~varf:(fun v -> v)
-      (Var.Set.empty, p.decl_ctx)
-      (* TODO: handle_default and handle_default_opt are now operators and not
-         variables. *)
-      p.code_items
+      Var.Set.empty p.code_items
+  in
+  (* Now we need to further tweak [decl_ctx] because some of the user-defined
+     types can have closures in them and these closured might have changed type.
+     So we reset them to [TAny] in the hopes that the transformation applied
+     will not yield to type unification conflicts. Indeed, consider the
+     following closure: [let f = if ... then fun v -> x + v else fun v -> v]. To
+     be typed correctly once converted, this closure needs an existential type
+     but the Catala typechecker doesn't have them. However, this kind of type
+     conflict is difficult to produce using the Catala surface language: it can
+     only happen if you store a closure which is the output of a scope inside a
+     user-defined data structure, and if you do it in two different places in
+     the code with two closures that don't have the same capture footprint. *)
```

Comment:
This comment was out of date, changing it. But to answer you, yes there's a risk that too many [TAny]s would prevent the typechecker to infer types completely ; empirically that hasn't been the case in all the tests I could see. And even if it were the case (the typechecker leaving a [TAny] after its pass) it would just mean that this type is unconstrained, so we can pick what we want for it (for instance [TUnit]).